contracts/contracts/rem.swap/src/rem.swap.cpp

```diff
@@ -0,0 +1,307 @@
+
+        std::vector<char> swap_payload( swap_key.data.begin(), swap_key.data.end() );
+        string payload_str = trx_id + chain_id + eth_address + quantity.to_string()
+                             + std::to_string( swap_init_timepoint.sec_since_epoch() );
+        std::copy( payload_str.begin(), payload_str.end(), back_inserter(swap_payload));
+
+        checksum256 swap_hash = hash( swap_payload.data() );
+        auto swap_hash_idx = swap_table.get_index<"byhash"_n>();
+        const auto swap_hash_itr = swap_hash_idx.find( swap_hash );
+        auto level = permission_level( user, "active"_n );
+
+        check( time_point_sec(current_time_point()) < swap_init_timepoint + swap_lifetime, "swap lifetime expired" );
+        cleanup_expired_swaps();
+        if ( swap_hash_itr == swap_hash_idx.end() ) {
+            swap_table.emplace(user, [&]( auto& s ) {
+                s.key = swap_table.available_primary_key();
+                s.trx_id = trx_id;
+                s.swap_hash = swap_hash;
+                s.swap_init_time = swap_init_timepoint;
+                s.status = static_cast<uint8_t>(swap_status::INITIALIZED);
+            });
+
+//            if ( is_block_producer(user) ) {
+//                swap_table.modify( *swap_hash_itr, user, [&]( auto& s ) {
+//                    s.provided_approvals.push_back( approval{ level, current_time_point() } );
+//                });
+//            }
+        }
+        else {
+//        check( is_block_producer(user), "authorization error" );
+            check( swap_hash_itr->status != static_cast<uint8_t>(swap_status::CANCELED), "swap already canceled" );
+            check( swap_hash_itr->status != static_cast<uint8_t>(swap_status::FINISHED), "swap already finished" );
+
+            const std::vector<approval>& approvals = swap_hash_itr->provided_approvals;
+            bool is_already_approval = std::find_if( approvals.begin(),
+                                                     approvals.end(),
+                                                     [&](const approval& a)
+                                                     { return a.level == level; }
+                                                     ) == approvals.end();
+
+            check ( is_already_approval, "approval already exist" );
+
+            swap_table.modify( *swap_hash_itr, user, [&]( auto& s ) {
+                s.provided_approvals.push_back( approval{ level, current_time_point() } );
+            });
+        }
+    }
+
+    // Debug function
+    void swap::cleartable( const name& user ) {
+        require_auth( user );
+
+        swap_index swap_table( _self, get_first_receiver().value );
+
+        for (auto _table_itr = swap_table.begin(); _table_itr != swap_table.end();) {
+            _table_itr = swap_table.erase(_table_itr);
+        }
+    }
+
+    void swap::finish( const name& user, const name& receiver, const string& trx_id,
```

Comment:
why do you need user variable here? You should sign account data - receiver's name, active and owner public keys (for new account)